### PR TITLE
:sparkles: (go/v4):  Add Hub and Spoke for conversion webhooks

### PR DESCRIPTION
**Pull request by camilamacedo86**
This PR introduces support for scaffolding a conversion webhook with a hub-and-spoke model in Kubebuilder Go plugin (v4). The feature allows defining one hub version with one or many spoke versions for seamless conversion between CustomResourceDefinition (CRD) API versions.

### How It Will Work:

Users can:
```
kubebuilder create webhook --group example.com --version v1 --kind MyKind --conversion --spoke v2
```

OR 

```
kubebuilder create webhook --group example.com --version v1 --kind MyKind --conversion --spoke v2,v3
```

Closes; https://github.com/kubernetes-sigs/kubebuilder/issues/2589

### CODE REVIEW

File: pkg/plugins/golang/v4/webhook.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v4/webhook.go:55]: Change here 
The flag would be better as spoke 
So it is more clear.

**Comment by camilamacedo86** [comment]
/hold 

Just to not merge before we do a patch release; https://github.com/kubernetes-sigs/kubebuilder/issues/4284

**Comment by camilamacedo86** [comment]
It is important because we have been adding e2e tests to ensure all features and options, including scaffolds and examples, as much as possible. The latest piece missing is the webhook conversion, which is not fully implemented yet but will be once we merge the PR: https://github.com/kubernetes-sigs/kubebuilder/pull/4254

File: pkg/model/resource/webhooks.go
Comment (varshaprasad96) [inline pkg/model/resource/webhooks.go:37]: `Spoke` can be multiple, so this should probably be `[]string`. Or do we want the user to execute this command multiple times for each spoke version? One issue would be that we need to ensure that `hub` is not scaffolded again as it may interfere with the already generated code.

File: pkg/plugins/golang/v4/scaffolds/internal/templates/webhooks/webhook.go
Comment (varshaprasad96) [inline pkg/plugins/golang/v4/scaffolds/internal/templates/webhooks/webhook.go:84]: Is this to enable executing conversion command multiple times? Since this template covers both validation and defaulting - there could be issues with just skipping file generation. Do we want to add print statements?

File: pkg/model/resource/webhooks.go
Comment (varshaprasad96) [inline pkg/model/resource/webhooks.go:84]: Since we are supporting only one `hub` - is there a way we can validate, where hub is not being changed. Or the spoke version was not previously declared as a `hub`. Or would this error occur while trying to scaffold the markers?

File: pkg/cli/alpha/internal/generate.go
Comment (varshaprasad96) [inline pkg/cli/alpha/internal/generate.go:364]: Probably add validation that `--conversion` needs to be set, to have `--spoke` version? Wondering if we need `--conversion` in that case? Can it be automatically considered true if user specifies `--spoke`?

File: docs/book/src/multiversion-tutorial/conversion.md
Comment (varshaprasad96) [inline docs/book/src/multiversion-tutorial/conversion.md:9]:
```suggestion
kubebuilder create webhook --group batch --version v1 --kind CronJob --conversion --spoke v2
```

File: pkg/cli/alpha/internal/generate.go
Comment (varshaprasad96) [reply pkg/cli/alpha/internal/generate.go:364]: Actually, rethinking about this - iiuc it is:
1. if `conversion` is specified - the existing functionality remains.
2. if `spoke` is specified - then `Hub` and `Spoke` interfaces are scaffolded.

In that case would it make sense to validate that `--conversion` is set if `--spoke` is specified.

File: pkg/plugins/golang/v4/scaffolds/internal/templates/api/spoke.go
Comment (varshaprasad96) [inline pkg/plugins/golang/v4/scaffolds/internal/templates/api/spoke.go:77]:
```suggestion
	log.Printf("Converting from %T to %T", src.APIVersion, dst.APIVersion)
```

IIUC it's the other way around - `ConverTo` converts to the hub version to store it in the API server. 
Eg: https://github.com/kubeflow/kubeflow/blob/ec82fbf58b79cff529d948b96e44ffd06bdfe679/components/notebook-controller/api/v1alpha1/notebook_conversion.go#L24

**Review by varshaprasad96 — COMMENTED** [review]
Thanks @camilamacedo86! Just a few questions on the `hub` and `spoke` specifications.

File: pkg/plugins/golang/v4/scaffolds/internal/templates/webhooks/webhook.go
Comment (camilamacedo86) [reply pkg/plugins/golang/v4/scaffolds/internal/templates/webhooks/webhook.go:84]: You are right here.
We have a terrible UX now to work with webhooks since is not possible create them without force as a follow up.
we need to fix that as well:  https://github.com/kubernetes-sigs/kubebuilder/issues/4146

I will think a solution here for we not be blocked by: https://github.com/kubernetes-sigs/kubebuilder/issues/4146
until we have this one.

File: pkg/cli/alpha/internal/generate.go
Comment (camilamacedo86) [reply pkg/cli/alpha/internal/generate.go:371]: If we do not inform the spoke that will fails because we will need to have that right.
Without the spoke we cannot implement it

File: pkg/model/resource/webhooks.go
Comment (camilamacedo86) [reply pkg/model/resource/webhooks.go:84]: I changed for []string
but maybe the right thing here would be have a List of GKV

File: docs/book/src/multiversion-tutorial/testdata/project/PROJECT
Comment (camilamacedo86) [inline docs/book/src/multiversion-tutorial/testdata/project/PROJECT:25]: @varshaprasad96 

Would it make sense to have Many spokes?
If I want to convert my V1 to V2
Why would I want to convert my V1 to V2 and v3 at the same time?

File: pkg/model/resource/webhooks.go
Comment (varshaprasad96) [inline pkg/model/resource/webhooks.go:105]: Is there a way to find if the provided spoke version is not one of hub? Given we are not storing the hub version in `Webhooks` struct? Or are we catching that error before?

**Review by varshaprasad96 — APPROVED** [review]
Thanks @camilamacedo86! Tested this out, and this looks good!
/lgtm
/approve

File: pkg/model/resource/webhooks.go
Comment (varshaprasad96) [reply pkg/model/resource/webhooks.go:105]: Sorry, just tested this out. Looks like we are catching it. Pls ignore.

**Pull request merged**
